pkg/daemon/server/service/pipeline_watermark_query.go

```diff
@@ -53,24 +52,39 @@ func newVertexWatermarkFetcher(pipeline *v1alpha1.Pipeline) *watermarkFetchers {
 			edge := pipeline.GetFromEdges(vertex.Name)[0]
 			fromBufferName = v1alpha1.GenerateEdgeBufferName(pipeline.Namespace, pipelineName, edge.From, edge.To)
 		}
-		hbBucket := isbsvc.JetStreamProcessorBucket(pipelineName, fromBufferName)
-		hbWatch, err := jetstream.NewKVJetStreamKVWatch(ctx, pipelineName, hbBucket, clients.NewInClusterJetStreamClient())
-		if err != nil {
-			log.Fatalw("JetStreamKVWatch failed", zap.String("HeartbeatBucket", hbBucket), zap.Error(err))
-		}
-		otBucket := isbsvc.JetStreamOTBucket(pipelineName, fromBufferName)
-		otWatch, err := jetstream.NewKVJetStreamKVWatch(ctx, pipelineName, otBucket, clients.NewInClusterJetStreamClient())
-		if err != nil {
-			log.Fatalw("JetStreamKVWatch failed", zap.String("OTBucket", otBucket), zap.Error(err))
+		// Adding an extra entry for the sink vertex to check the watermark value progressed out of the vertex.
+		// Can be checked by querying sinkName_SINK in the service
+		if vertex.Sink != nil {
+			toBufferName := v1alpha1.GenerateSinkBufferName(pipeline.Namespace, pipelineName, vertex.Name)
+			fetchWatermark := createWatermarkFetcher(ctx, pipelineName, toBufferName, vertex.Name)
+			sinkVertex := vertex.Name + "_SINK"
+			vertexWmMap[sinkVertex] = fetchWatermark
 		}
-		var fetchWmWatchers = generic.BuildFetchWMWatchers(hbWatch, otWatch)
-		fetchWatermark := generic.NewGenericFetch(ctx, vertex.Name, fetchWmWatchers)
+		fetchWatermark := createWatermarkFetcher(ctx, pipelineName, fromBufferName, vertex.Name)
 		vertexWmMap[vertex.Name] = fetchWatermark
 	}
 	wmFetcher.fetchMap = vertexWmMap
 	return wmFetcher
 }
 
+func createWatermarkFetcher(ctx context.Context, pipelineName string, fromBufferName string, vertexName string) *generic.GenericFetch {
+	log := logging.FromContext(ctx)
+	hbBucket := isbsvc.JetStreamProcessorBucket(pipelineName, fromBufferName)
+	hbWatch, err := jetstream.NewKVJetStreamKVWatch(ctx, pipelineName, hbBucket, clients.NewInClusterJetStreamClient())
+	if err != nil {
+		log.Fatalw("JetStreamKVWatch failed", zap.String("HeartbeatBucket", hbBucket), zap.Error(err))
```

Comment:
Please return error and let the invoker to decide how to handle the error instead of letting it panic here.